architecture/2025-01-serverless-feature-spec.md

```diff
@@ -0,0 +1,534 @@
+
+## Current state
+<!-- If we already have some ongoing investment in this area, summarize the current state and point to any relevant documents. -->
+There has been a proof-of-concept for ACI support in Radius built on top of the ACI nGroups features. Here is a demo of the prototype that has been created so far: https://youtu.be/NMNZE22nSQI?si=Dq7Q5WVKgHularsO&t=1201
+
+## Detailed user experience
+ <!-- <List of steps the user goes through from the start to the end of the scenario to provide more detailed view of exactly what the user is able to do given the new capabilities>  -->
+<!-- Step 1
+Step 2
+… -->
+
+### Step 1: Define and deploy a Radius Environment for serverless
+
+The user defines a new Radius Environment for serverless compute by creating a new Environment definition file (e.g. `env.bicep`) and specifying the necessary settings for the serverless platform before deploying the environment using `rad deploy env.bicep`.
+
+```diff
+resource environment 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'myenv'
+  properties: {
+    compute: {
++     // Required. The kind of container runtime to use, 
++     //  e.g. 'aci' for Azure Container Instances and 'ecs' for AWS Elastic Container Service
++     kind: 'aci'
++     // This is currently required, but may be made optional if 
++     //  it doesn't apply to all compute platforms
++     namespace: 'default'
++     // This was introduced as a part of the ACI prototype, but may be changed 
++     //  depending on the implementation, need to figure out if namespace and 
++     //  resourceGroup should be combined into a single more generic property
++     resourceGroup: '/subscriptions/.../resourceGroups/myrg'
+    }
++   // one provider per environment    
++   providers: {
++     // if the compute platform for this environment is ACI, then the provider must be Azure
++     azure: {
++       scope: '/subscriptions/.../resourceGroups/myrg'
++     }
++     // if the compute platform for this environment is ECS, then the provider must be AWS
++     aws: {
++       scope: '/planes/aws/aws/accounts/${account}/regions/${region}'
++     }
++   }
++   extensions: [ // these are all currently Kubernetes-centric, will need to update for serverless
+      {
+        kind: 'kubernetesMetadata'
+        labels: {
+          'team.contact.name': 'frontend'
+        }
++       // Add other serverless platform-specific extensions here
+      }
+    ]
+  }
+}
+```
+
+> Note: The `providers` property in the environment is necessary to specify the scope of the environment and is different from the provider credentials that were registered using `rad init`. The provider credentials are used to authenticate with the underlying compute provider, while the `providers` environment property is used to specify the scope of the environment. Thus, we might consider renaming the Environment `providers` property to avoid confusion.
```

Comment:
Are we planning to restrict credentials to map 1-1 with the compute platform as well?

Reply:
I don't think we'd add that restriction since the rad credentials can apply to things beyond the compute.